src/file-upload/dropzone.tsx

```diff
@@ -0,0 +1,113 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React, { useEffect, useRef, useState } from 'react';
+import styles from './styles.css.js';
+import clsx from 'clsx';
+import InternalIcon from '../icon/internal';
+import InternalBox from '../box/internal';
+
+interface DropzoneProps {
+  onChange: (files: File[]) => void;
+  children: React.ReactNode;
+}
+
+export function useDropzoneVisible() {
+  const [isDropzoneVisible, setDropzoneVisible] = useState(false);
+
+  // Registering global drag events listeners.
+  useEffect(() => {
+    // The timer helps avoiding dropzone blinking.
+    let dragTimer: null | ReturnType<typeof setTimeout> = null;
+
+    // The file-upload dropzone is shown when the user drags files over the to the browser.
+    const onDocumentDragOver = (event: DragEvent) => {
+      event.preventDefault();
+
+      if (event.dataTransfer && event.dataTransfer.types && event.dataTransfer.types.indexOf('Files') !== -1) {
+        setDropzoneVisible(true);
+        dragTimer && clearTimeout(dragTimer);
+      }
+    };
+
+    // When the files are no longer dragged over the browser the state must be reset.
+    const onDocumentDragLeave = (event: DragEvent) => {
+      event.preventDefault();
+
+      dragTimer = setTimeout(() => setDropzoneVisible(false), 25);
+    };
+
+    // If the files were dropped the state must be reset.
+    const onDocumentDrop = (event: DragEvent) => {
+      event.preventDefault();
+
+      dragTimer = setTimeout(() => setDropzoneVisible(false), 25);
+    };
+
+    document.addEventListener('dragover', onDocumentDragOver, false);
+    document.addEventListener('dragleave', onDocumentDragLeave, false);
+    document.addEventListener('drop', onDocumentDrop, false);
+
+    return () => {
+      dragTimer && clearTimeout(dragTimer);
+      document.removeEventListener('dragover', onDocumentDragOver);
+      document.removeEventListener('dragleave', onDocumentDragLeave);
+      document.removeEventListener('drop', onDocumentDrop);
+    };
+  }, []);
+
+  return isDropzoneVisible;
+}
+
+export function Dropzone({ onChange, children }: DropzoneProps) {
+  const dropzoneRef = useRef<HTMLDivElement>(null);
+  const [isDropzoneHovered, setDropzoneHovered] = useState(false);
+
+  const onDragOver = (event: React.DragEvent) => {
+    event.preventDefault();
+    event.stopPropagation();
```

Comment:
Why do we need to stop propagation?

Reply:
I added it when experimenting with dataTransfer but it is not really needed. I'll remove it.